src/cli/pagerank.js

```diff
@@ -0,0 +1,202 @@
+ * which is responsible for saving the resultant PagerankGraph to disk.
+ */
+export function makePagerankCommand(
+  loader: (RepoId) => Promise<LoadGraphResult>,
+  pagerankRunner: (Graph) => Promise<PagerankGraph>,
+  saver: (RepoId, PagerankGraph) => Promise<void>
+): Command {
+  return async function pagerank(args, std) {
+    let repoId: RepoId | null = null;
+    for (let i = 0; i < args.length; i++) {
+      switch (args[i]) {
+        case "--help": {
+          usage(std.out);
+          return 0;
+        }
+        default: {
+          if (repoId != null)
+            return die(std, "multiple repository IDs provided");
+          // Should be a repository.
+          repoId = stringToRepoId(args[i]);
+          break;
+        }
+      }
+    }
+
+    if (repoId == null) {
+      return die(std, "no repository ID provided");
+    }
+
+    const result: LoadGraphResult = await loader(repoId);
+
+    switch (result.status) {
+      case "REPO_NOT_LOADED": {
+        const repoIdStr = repoIdToString(repoId);
+        std.err(`fatal: repository ID ${repoIdStr} not loaded`);
+        std.err(`Try running \`sourcecred load ${repoIdStr}\` first.`);
+        return 1;
+      }
+      case "PLUGIN_FAILURE": {
+        std.err(
+          `fatal: plugin "${result.pluginName}" errored: ${
+            result.error.message
+          }`
+        );
+        return 1;
+      }
+      case "SUCCESS": {
+        const pagerankGraph = await pagerankRunner(result.graph);
+        await saver(repoId, pagerankGraph);
+        return 0;
+      }
+      // istanbul ignore next: unreachable per Flow
+      default: {
+        std.err(`Unexpected status: ${(result.status: empty)}`);
+        return 1;
+      }
+    }
+  };
+}
+
+export async function runPagerank(
+  weights: WeightedTypes,
+  graph: Graph
+): Promise<PagerankGraph> {
+  const evaluator = weightsToEdgeEvaluator(weights);
+  const pagerankGraph = new PagerankGraph(
+    graph,
+    evaluator,
+    DEFAULT_SYNTHETIC_LOOP_WEIGHT
+  );
+  await pagerankGraph.runPagerank({
+    maxIterations: DEFAULT_MAX_ITERATIONS,
+    convergenceThreshold: DEFAULT_CONVERGENCE_THRESHOLD,
+  });
+  return pagerankGraph;
+}
+
+export async function savePagerankGraph(
+  directory: string,
+  repoId: RepoId,
+  pg: PagerankGraph
+): Promise<void> {
+  const pgJSON = pg.toJSON();
+  const pgDir = path.join(directory, "data", repoIdToString(repoId));
+  await fs.ensureDir(pgDir);
+  const pgFile = path.join(pgDir, "pagerankGraph.json");
+  await fs.writeFile(pgFile, stringify(pgJSON));
+}
+
+function weightsForAdapters(
+  adapters: $ReadOnlyArray<IAnalysisAdapter>
+): WeightedTypes {
+  const declarations = adapters.map((a) => a.declaration());
+  return combineWeights(declarations.map(defaultWeightsForDeclaration));
+}
+
+export const defaultAdapters = () => [
+  new GithubAnalysisAdapter(),
+  new GitAnalysisAdapter(),
+  new FallbackAdapter(),
```

Comment:
The astute reviewer will note that this is a little different from the other commands, in that we include the FallbackAdapter. This is needed for tests to pass, as otherwise the pagerank command would fail if a node was included in the graph that no plugin accounted for (as would be the case, for example, if the GitHub plugin were run on its own, as it creates git nodes without providing corresponding NodeTypes). Handling this case is why we have a fallbackAdapter to begin with.

In a future-but-soon PR, I plan to refactor the list of default plugins out of the individual commands and into a shared file, along with an interface that makes it clear what a plugin needs to package.